yktracec/src/jitmodbuilder.cc

```diff
@@ -1228,7 +1246,16 @@ class JITModBuilder {
               assert(Idx + 1 < InpTrace.Length());
               TraceLoc MaybeNextIB = InpTrace[Idx + 1];
               if (const IRBlock *NextIB = MaybeNextIB.getMappedBlock()) {
-                CF = AOTMod->getFunction(NextIB->FuncName);
+                auto isNextBlockNotEntry = NextIB->BBIdx > 0;
+                if (IsSWTrace && CI->isIndirectCall() && isNextBlockNotEntry) {
```

Comment:
I have been chewing on this part of the PR for a while and I think there might be an issue with it.

Imagine we have this program

```
fn f(x: Fn(...)) {
   x();
}

fn g() {
  ...
}
```

and we're tracing `f`. It calls a function pointed to by `x`in a way that we can't know statically. Imagine that we call some code which contains `f(&g)` --- then the trace would look something like:

```
start_f()
indirect_call(*x)
start_g()
...
```

and would be handled by line 1250 in your PR. But imagine that in another place I have code which contains `f(&h)` where `h` is unmappable and looks like:

```
fn h() {
  g()
}
```

This will also create a trace:

```
start_f()
indirect_call(*x)
start_g()
...
```

In other words, I don't *think* we can tell from the software trace whether some unknown code was called as part of the indirect function call as we get the same trace in either case. Warning: I might be wrong!

If I'm not wrong then we need to think of solutions. Two jump out at me:

1. Throw away any trace which contains an indirect function call. This is easy, but obviously unsatisfying.
2. Have the software tracer record the *value* of the function pointer. Then jitmodbuilder (or, probably, the trace mapper) can say:
      a. The pointer is to a function with IR and we can inline it easily.
      b. The pointer is to a function without IR and we can't inline it easily.

CCing @vext01 and @ptersilie  who might have additional thoughts.

Reply:
If what you are saying is that we can't tell the difference betwen `g` being called directly via an indirect call or indirectly via another unmappable function, then yes that is correct. However, since an indirect call starts outlining, both cases would be handled the same:
1) See indirect call. Start outlining
2) See entry block of g: push callstack
3) See return from g: pop callstack. Now we stop outlining since we've reached f again
4) Continue copying instructions as normal.

Reply:
@ptersilie Good point, that is an option 3 that I hadn't considered. It means we lose the ability to inline some safe things, but does mean we don't inline unsafe things. So I think my option 1 is pointless for this PR: we should do option 3 (i.e. what you've suggested) in this PR (I guess that *is* what the PR does already?). And then we can try my option 2 in a later PR.

Reply:
I just had another look at this code (particularly line 1250). So what this does for the hwt is to peek ahead in the trace and check if the next block is a mappable block. This way we know the indirect call is to something we can inline (if the next block is unmappable then we start outlining). The change in this commit was added to make swt do the same. However, thanks to @ltratt's example I now realise this is not correct. Peek ahead in the software tracer might give you a mappable entry block, however we don't know if there was an umappable call in between. So what that means for this PR is that we always have to outline indirect calls in the swt (at least for the time being). To do so I think we just need to turn that above condition into `if (IsSWTrace)`.

Reply:
updated 👉 b7237000e85848d3b498eff3bf1ca9425236359c

Reply:
Works for me.